File: payroll-api/src/main/java/com/company/payroll/service/impl/StaffMiscellaneousServiceImpl.java
package com.company.payroll.service.impl;

import java.io.FileNotFoundException;
import java.math.BigDecimal;
import java.nio.file.Paths;
import java.util.List;
import java.util.Objects;
import java.util.Optional;

import com.company.payroll.model.*;
import com.company.payroll.util.FileUtils;
import org.apache.tomcat.util.http.fileupload.InvalidFileNameException;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.core.io.Resource;
import org.springframework.stereotype.Service;

import com.company.payroll.mapper.StaffDetailsMapper;
import com.company.payroll.mapper.ManagerMapper;
import com.company.payroll.mapper.StaffPromotionMapper;
import com.company.payroll.mapper.StaffResignationMapper;
import com.company.payroll.mapper.StaffSalaryMapper;
import com.company.payroll.mapper.StaffTrainingMapper;
import com.company.payroll.model.StaffSalary;
import com.company.payroll.service.StaffMiscellaneousService;
import com.github.pagehelper.PageHelper;
import com.github.pagehelper.PageInfo;
import org.springframework.web.multipart.MultipartFile;

@Service
public class StaffMiscellaneousServiceImpl implements StaffMiscellaneousService {

	@Autowired
	private StaffDetailsMapper staffDetailsMapper;

	@Autowired
	private StaffPromotionMapper staffPromotionMapper;

	@Autowired
	private StaffSalaryMapper staffSalaryMapper;
	
	@Autowired
	private StaffResignationMapper staffResignationMapper;
	
	@Autowired
	private StaffTrainingMapper staffTrainingMapper;
	
	@Autowired
	private FileUtils fileUtils;
	
	@Override
	public Integer deletePromotion(Integer pId) {
		return staffPromotionMapper.deleteByPrimaryKey(pId);
	}

	@Override
	public Integer deleteResignation(Integer resignId) {
		int row = 0;

		Optional<StaffResignation> staffResignation = Optional.ofNullable(staffResignationMapper.selectByPrimaryKey(resignId));
		if(staffResignation.isPresent()) {
			boolean bool = fileUtils.delete(Paths.get(staffResignation.get().getFilePath()));
			if(bool) {
				row += staffDetailsMapper.deleteByPrimaryKey(resignId);
			}
		} else {
			return null;
		}

		return row;
	}

	@Override
	public Integer deleteTraining(Integer tId) {
		return staffTrainingMapper.deleteByPrimaryKey(tId);
	}

	@Override
	public Optional<StaffPromotion> findPromotionById(Integer pId) {
		return Optional.ofNullable(staffPromotionMapper.selectByPrimaryKey(pId));
	}

	@Override
	public Optional<StaffResignation> findResignationById(Integer resignId) {
		return Optional.ofNullable(staffResignationMapper.selectByPrimaryKey(resignId));
	}

	@Override
	public Optional<List<StaffTraining>> findTrainingByEId(int eId) {
		return Optional.empty();
	}

	@Override
	public Optional<List<StaffTraining>> findTrainingByMId(int mId) {
		return Optional.empty();
	}

	@Override
	public List<StaffTraining> listTrainingByStaffId(Integer staffId) {
		return staffTrainingMapper.selectListByStaffId(staffId);
	}

	@Override
	public Optional<StaffTraining> findTrainingById(Integer tId) {
		return Optional.ofNullable(staffTrainingMapper.selectByPrimaryKey(tId));
	}

	@Override
	public StaffPromotion insertPromotion(StaffPromotion staffPromotion) {
		StaffDetails staffDetails = staffDetailsMapper.selectByPrimaryKey(staffPromotion.getStaffId());
		staffDetails.setTitleNo(staffPromotion.getTitleNo());
		staffDetailsMapper.updateByPrimaryKeySelective(staffDetails);

		StaffSalary staffSalary = staffSalaryMapper.selectByPrimaryKey(staffDetails.getSId());
		staffSalary.setMonthlySalary(staffPromotion.getPromoteSalary());
		staffSalary.setAnnualSalary(staffSalary.getMonthlySalary().multiply(BigDecimal.valueOf(13)));
		staffSalary.setDateUpdate(staffPromotion.getPromoteDate());
		staffSalaryMapper.updateByPrimaryKeySelective(staffSalary);

		staffPromotionMapper.insertSelective(staffPromotion);

		return staffPromotion;
	}

	@Override
	public StaffResignation insertResignation(StaffResignation resign) {
		return null;
	}

	@Override
	public StaffResignation insertResignation(MultipartFile attachment, StaffResignation staffResignation) {
		StaffDetails staffDetails = staffDetailsMapper.selectByPrimaryKey(staffResignation.getStaffId());
		staffDetails.setDateResign(staffResignation.getResignDate());
		staffDetailsMapper.updateByPrimaryKeySelective(staffDetails);

		String filePath = "/staffs/resign/" + staffResignation.getStaffId().toString();

		if(Objects.equals(attachment.getContentType(), "application/msword") || Objects.equals(attachment.getContentType(), "application/pdf")
				||Objects.equals(attachment.getContentType(), "application/wps-office.doc") || Objects.equals(attachment.getContentType(), "application/wps-office.docx")) {
			staffResignation.setFileName(attachment.getOriginalFilename());
			staffResignation.setFileSize(attachment.getSize());
			staffResignation.setFilePath(fileUtils.fileUpload(attachment, filePath));

			staffResignationMapper.insertSelective(staffResignation);
		} else {
			throw new InvalidFileNameException(attachment.getOriginalFilename(), "Invalid file format");
		}

		return staffResignation;
	}

	@Override
	public StaffTraining insertTraining(StaffTraining staffTraining) {
		staffTrainingMapper.insertSelective(staffTraining);
		return staffTraining;
	}

	@Override
	public PageInfo<StaffPromotion> listPromotion(int page, int offset) {
		PageHelper.startPage(page, offset);
		return new PageInfo<StaffPromotion>(staffPromotionMapper.selectList());
	}

	@Override
	public PageInfo<StaffPromotion> listPromotionByEId(int page, int offset, int eId) {
		return null;
	}

	@Override
	public PageInfo<StaffPromotion> listPromotionByStaffId(int page, int offset, Integer staffId) {
		PageHelper.startPage(page, offset);
		return new PageInfo<StaffPromotion>(staffPromotionMapper.selectListByStaffId(staffId));
	}

	@Override
	public PageInfo<StaffResignation> listResignation(int page, int offset) {
		PageHelper.startPage(page, offset);
		return new PageInfo<StaffResignation>(staffResignationMapper.selectList());
	}

	@Override
	public PageInfo<StaffTraining> listTraining(int page, int offset) {
		PageHelper.startPage(page, offset);
		return new PageInfo<StaffTraining>(staffTrainingMapper.selectList());
	}

	@Override
	public StaffPromotion updatePromotion(StaffPromotion staffPromotion) {
		StaffDetails staffDetails = staffDetailsMapper.selectByPrimaryKey(staffPromotion.getStaffId());
		staffDetails.setTitleNo(staffPromotion.getTitleNo());
		staffDetailsMapper.updateByPrimaryKeySelective(staffDetails);

		StaffSalary staffSalary = staffSalaryMapper.selectByPrimaryKey(staffDetails.getSId());
		staffSalary.setMonthlySalary(staffPromotion.getPromoteSalary());
		staffSalary.setAnnualSalary(staffSalary.getMonthlySalary().multiply(BigDecimal.valueOf(13)));
		staffSalary.setDateUpdate(staffPromotion.getPromoteDate());
		staffSalaryMapper.updateByPrimaryKeySelective(staffSalary);

		staffPromotionMapper.updateByPrimaryKeySelective(staffPromotion);

		return staffPromotion;
	}

	@Override
	public StaffResignation updateResignation(StaffResignation staffResignation) {
		StaffDetails staffDetails = staffDetailsMapper.selectByPrimaryKey(staffResignation.getStaffId());
		staffDetails.setDateResign(staffResignation.getResignDate());
		staffDetailsMapper.updateByPrimaryKeySelective(staffDetails);

		staffResignationMapper.updateByPrimaryKeySelective(staffResignation);
		return staffResignation;
	}

	@Override
	public StaffTraining updateTraining(StaffTraining staffTraining) {
		staffTrainingMapper.updateByPrimaryKeySelective(staffTraining);
		return staffTraining;
	}

	@Override
	public Resource downloadResignationAttachment(Integer resignId) {
		StaffResignation resignationDetails = staffResignationMapper.selectByPrimaryKey(resignId);

		return fileUtils.download(Paths.get(resignationDetails.getFilePath()));
	}

}

File: kotlin-payroll/src/main/kotlin/com/company/payroll/service/impl/StaffMiscellaneousServiceImpl.kt
package com.company.payroll.service.impl

import com.company.payroll.mapper.*
import com.company.payroll.model.StaffPromotion
import com.company.payroll.model.StaffResignation
import com.company.payroll.model.StaffTraining
import com.company.payroll.service.StaffMiscellaneousService
import com.company.payroll.util.FileUtils
import com.github.pagehelper.PageHelper
import com.github.pagehelper.PageInfo
import org.apache.tomcat.util.http.fileupload.InvalidFileNameException
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import org.springframework.web.multipart.MultipartFile
import java.math.BigDecimal
import java.nio.file.Paths

@Service
class StaffMiscellaneousServiceImpl(@Autowired private val staffPromotionMapper: StaffPromotionMapper,
                                    @Autowired private val staffDetailsMapper: StaffDetailsMapper,
                                    @Autowired private val staffSalaryMapper: StaffSalaryMapper,
                                    @Autowired private val staffResignationMapper: StaffResignationMapper,
                                    @Autowired private val fileUtils: FileUtils,
                                    @Autowired private val staffTrainingMapper: StaffTrainingMapper): StaffMiscellaneousService {
  override fun deletePromotion(pId: Int): Int {
    return staffPromotionMapper.deleteByPrimaryKey(pId)
  }

  override fun deleteResignation(resignId: Int): Int {
    var row = 0
    val resign = staffResignationMapper.selectByPrimaryKey(resignId)
    val bool = resign.filePath?.let { Paths.get(it) }?.let { fileUtils.delete(it) }

    if(bool == true) {
      row = staffResignationMapper.deleteByPrimaryKey(resignId)
    }

    return row
  }

  override fun deleteTraining(tId: Int): Int {
    return staffTrainingMapper.deleteByPrimaryKey(tId)
  }

  override fun findPromotionById(pId: Int): StaffPromotion {
    return staffPromotionMapper.selectByPrimaryKey(pId)
  }

  override fun findResignationById(resignId: Int): StaffResignation {
    return staffResignationMapper.selectByPrimaryKey(resignId)
  }

  override fun findTrainingById(tId: Int): StaffTraining {
    return staffTrainingMapper.selectByPrimaryKey(tId)
  }

  override fun insertPromotion(staffPromotion: StaffPromotion): StaffPromotion {
    val staffDetails = staffDetailsMapper.selectByPrimaryKey(staffPromotion.staffId)
    staffDetails.titleNo = staffPromotion.titleNo
    staffDetailsMapper.updateByPrimaryKeySelective(staffDetails)

    val staffSalary = staffSalaryMapper.selectByPrimaryKey(staffDetails.sId)
    staffSalary.monthlySalary = staffPromotion.promoteSalary
    staffSalary.annualSalary = staffSalary.monthlySalary.multiply(BigDecimal(13))
    staffSalary.dateUpdate = staffPromotion.promoteDate
    staffSalaryMapper.updateByPrimaryKeySelective(staffSalary)

    staffPromotionMapper.insertSelective(staffPromotion)

    return staffPromotion
  }

  override fun insertResignation(attachment: MultipartFile, staffResignation: StaffResignation): StaffResignation {
    val staffDetails = staffDetailsMapper.selectByPrimaryKey(staffResignation.staffId)
    staffDetails.dateResign = staffResignation.resignDate
    staffDetailsMapper.updateByPrimaryKeySelective(staffDetails)

    val filepath = "/staffs/resign/${staffResignation.staffId}"

    if (attachment.contentType == "application/msword" || attachment.contentType == "application/pdf" ||
        attachment.contentType == "application/wps-office.doc" || attachment.contentType == "application/wps-office.docx") {
      staffResignation.fileName = attachment.originalFilename
      staffResignation.fileSize = attachment.size
      staffResignation.filePath = fileUtils.fileUpload(attachment, filepath)

      staffResignationMapper.insertSelective(staffResignation)
    } else {
      throw InvalidFileNameException(attachment.name, "Invalid file format.")
    }

    return staffResignation
  }

  override fun insertTraining(staffTraining: StaffTraining): StaffTraining {
    staffTrainingMapper.insertSelective(staffTraining)
    return staffTraining
  }

  override fun listPromotion(page: Int, offset: Int): PageInfo<StaffPromotion> {
    PageHelper.startPage<StaffPromotion>(page, offset)
    return PageInfo<StaffPromotion>(staffPromotionMapper.selectList())
  }

  override fun listPromotionByStaffId(page: Int, offset: Int, staffId: Int): PageInfo<StaffPromotion> {
    PageHelper.startPage<StaffPromotion>(page, offset)
    return PageInfo<StaffPromotion>(staffPromotionMapper.selectListByStaffId(staffId))
  }

  override fun listResignation(page: Int, offset: Int): PageInfo<StaffResignation> {
    PageHelper.startPage<StaffResignation>(page, offset)
    return PageInfo<StaffResignation>(staffResignationMapper.selectList())
  }

  override fun listTraining(page: Int, offset: Int): PageInfo<StaffTraining> {
    PageHelper.startPage<StaffTraining>(page, offset)
    return PageInfo<StaffTraining>(staffTrainingMapper.selectList())
  }

  override fun listTrainingByStaffId(staffId: Int): List<StaffTraining> {
    return staffTrainingMapper.selectListByStaffId(staffId)
  }

  override fun updatePromotion(staffPromotion: StaffPromotion): StaffPromotion {
    val staffDetails = staffDetailsMapper.selectByPrimaryKey(staffPromotion.staffId)
    staffDetails.titleNo = staffPromotion.titleNo
    staffDetailsMapper.updateByPrimaryKeySelective(staffDetails)

    val staffSalary = staffSalaryMapper.selectByPrimaryKey(staffDetails.sId)
    staffSalary.monthlySalary = staffPromotion.promoteSalary
    staffSalary.annualSalary = staffSalary.monthlySalary.multiply(BigDecimal(13))
    staffSalary.dateUpdate = staffPromotion.promoteDate
    staffSalaryMapper.updateByPrimaryKeySelective(staffSalary)

    staffPromotionMapper.updateByPrimaryKeySelective(staffPromotion)

    return staffPromotion
  }

  override fun updateResignation(staffResignation: StaffResignation): StaffResignation {
    val staffDetails = staffDetailsMapper.selectByPrimaryKey(staffResignation.staffId)
    staffDetails.dateResign = staffResignation.resignDate
    staffDetailsMapper.updateByPrimaryKeySelective(staffDetails)

    staffResignationMapper.updateByPrimaryKeySelective(staffResignation)

    return staffResignation
  }

  override fun updateTraining(staffTraining: StaffTraining): StaffTraining {
    staffTrainingMapper.updateByPrimaryKeySelective(staffTraining)
    return staffTraining
  }
}